packages/tailwindcss/src/variants.ts

```diff
@@ -226,8 +226,49 @@ export class Variants {
     }
 
     let compareFn = this.compareFns.get(aOrder)
-    if (compareFn === undefined) return a.root < z.root ? -1 : 1
-    return compareFn(a, z)
+    if (compareFn !== undefined) return compareFn(a, z)
+
+    if (a.root !== z.root) return a.root < z.root ? -1 : 1
+
+    // SAFETY: variants a and z are of the same `kind` and can only be static or
+    // functional at this point
+
+    let aValue =
+      a.kind === 'static'
+        ? // Static variants don't have a value
+          null
+        : a.kind === 'compound'
+          ? // Compound variants never reach this point
+            null
+          : a.value
+
+    let zValue =
+      z.kind === 'static'
+        ? // Static variants don't have a value
+          null
+        : z.kind === 'compound'
+          ? // Compound variants never reach this point
+            null
+          : z.value
+
+    // Both variants are static, functional without a value, or functional with
+    // equal values
+    if (aValue === zValue) return 0
```

Comment:
Well one this was supposed to be `aValue.value` and `zValue.value` but upon further investigation this isn't actually observable _because of the way we sort valid utilities_.

Reply:
It _is_ observable when we sort this:
```ts
  let classLists = permute([
    //
    'data-[foo=bar baz]:flex',
    'data-[foo=bar_baz]:flex',
  ])
```

Notice the ` ` in there instead of a `_`. They both parse to the same thing but sort differently if this check isn't there. But… ` ` isn't a valid character in a class so it can't actually happen in practice.

Reply:
Okay I removed the code. I popped in a few `SAFETY` blocks explaining why we don't need equality checks and always returning 1 or -1 is fine. (normally this is not okay because not returning 0 for equal values results in an unstable sort)